### DISCUSSION TITLE
collision detection when creating IDs

### DISCUSSION THREAD

**Discussion by wt** [maxcountryman/tower-sessions#169] [Q&A]
I haven't found any collision specific detection when creating a new session record. I do see that a store to the session store can fail, which results in an internal server error response. This is probably reasonable since a 128 bit random value like Id shouldn't collide very often.

However, I was wondering if we could have a new session_store::Error type that indicates a duplicate id was attempted to be stored. Then, maybe react to that error with a retry on generating a new session id, ideally logging whenever a collision occurs. This overhead would only be incurred when creating a new session Id and would prevent an unfortunate internal server error.

**Comment by maxcountryman** [comment]
> However, I was wondering if we could have a new session_store::Error type that indicates a duplicate id was attempted to be stored.

This seems reasonable to me.

It's a bit unfortunate that this is a breaking change that requires store implementations to handle a new variant.

**Comment by wt** [comment]
What if we do it in two steps? The next minor release of tower-sessions could just include the new enum variant (this would be a major release for the tower-sessions-core), but not change the error handling by default (thus still produce an internal server error). This would be compilable for all store implementations that don't match on the Error enum variants. I can't think of a valid use case for that so I think it might be reasonable to do that. Also, stores would be allowed to start returning the new error variant. If a store does that and is used with the older version, they would not get retries and would still get an internal server error. Apps could also opt into the retry by constructing the ServiceManager with a different constructor (and a corresponding new constructor in the ServiceManagerLayer that allows opting into the retry functionality. Something like this:

```
let session_store = MemoryStore::default();
// None for second arg picks default number of retries Some(i32) could explicitly pick the number
SessionManagerLayer::new_with_id_collision_retries(session_store, None) 
    .with_same_site(SameSite::Lax)
    .with_expiry(Expiry::OnInactivity(DEFAULT_SESSION_TIMEOUT))
```

For the major version, we could change the ServieManager and ServiceManagerLayers to use builders for construction and have a builder method that turns on the functionality, like this:

```
let session_store = MemoryStore::default();
SessionManagerLayer::builder()
    .with_store(session_store)
     // Some(i16) explicitly chooses number of retries; None picks default number of retries
    .with_id_collision_retries(Some(3))
    .build()
    .with_same_site(SameSite::Lax)
    .with_expiry(Expiry::OnInactivity(DEFAULT_SESSION_TIMEOUT))
```

Alternatively, we could move all the builder type methods in SessionManagerLayer to the builder.

Or...we could technically add a method to the SessionConfig to enable retries and add a method to the SessionManager to enable the retries and make the new constructor or add a new to SessionManagerLayer to enable the retries. I think this is the ugliest solution because it's not clear to me that the config for the retries should live in the SessionConfig.

I get that adding the enum variant by itself is technically a breaking change, but this seems pretty low risk for causing issues with SessionStores.

Obviously, we could also just implement whatever design with a new major version for tower-sessions. If we have to do that, do you have a timeline for the next major release?

Do you have any opinions on these options?

**Reply by maxcountryman** [reply]
> If we have to do that, do you have a timeline for the next major release?

We can cut a new release as soon as we've merged in the changes. There's nothing stopping us there.

**Comment by wt** [comment]
Also, if we must do a major revision, maybe we can redesign the Error a bit to hide some of the details as indicated in the docs for the thiserror crate [here](https://docs.rs/thiserror/latest/thiserror/#:~:text=Another%20use%20case,Debug%29%5D%0Aenum%20ErrorRepr%20%7B%0A%20%20%20%20...%0A%7D).

Maybe we should do something like this:
```
#[derive(thiserror::Error, Debug)]
#[error(transparent)]
pub struct Error(#[from] ErrorRepr);

impl Error {
    fn is_retryable(&self) -> bool {
        match *self {
            Encode(_) => false,
            Decode(_) => false,
            Backend(retryable, _) => retryable,
        }
    }
}

#[derive(thiserror::Error, Debug)]
struct BackendError {
    retryable: bool,
    msg: String
}

// Private and free to change across minor version of the crate.
#[derive(thiserror::Error, Debug)]
enum ErrorRepr {
    #[error("Encoding failed with: {0}")]
    Encode(String),

    #[error("Decoding failed with: {0}")]
    Decode(String),

    #[error("{0.msg}")]
    Backend(BackendError),
}
```

**Reply by maxcountryman** [reply]
I don't see a way around the version bump since it's a public type and adding a variant is a breaking change re an exhaustive type.

Still I'm not opposed to overhauling the error type and accounting for this specific scenario. With that said, collisions are low probability, "It’s akin to the chance of randomly picking a specific atom out of all the atoms in the observable universe." So is this worth the churn?

**Reply by wt** [reply]
PHP has id collision logic like this. That's what got me curious about this.

I would expect this logic to only be exercised by extremely large apps (birthday paradox) or by bugs in stores or id generation (like bad random data like in the older /dev/urandom days on Linux or maybe in embedded systems). I think both of these are worth mitigating in a way that the app owner can see it's happening without exposing it to users (via a 5xx response).

**Comment by maxcountryman** [comment]
I'm thinking a bit more here and (apologies I'm not at the computer so I may be misremembering) isn't it the case that a preexisting ID would not result in an error in the first place?

Consider that we are either saving, loading, or deleting a record--none of these actions encode the fact that a record should not already exist. What I believe this means is a collision wouldn't result in an error, it would just operate over the associated session.

Maybe I've missed something here so again please jump in and correct me.

If I'm right tho this is a serious problem (albeit one still as unlikely as we've previously pointed out) because it means a collision could give you someone else's session (!). To address this, we might need a new session store method, like "create", to ensure that we can identify when a session should not already exist.

**Reply by wt** [reply]
It depends. If I were implementing a store over an sql DB and tried to store the new session, I could raise an error at that point. Doing so would cause the current logic as I read it to return no set-cookie header and generate an internal service error. See code [here](https://github.com/maxcountryman/tower-sessions/blob/main/src/service.rs#L168-L175).

Assuming no bug in the store, I think that that is enough to stop a hijack if the session id's collide.

**Reply by wt** [reply]
Oh, I guess it would not be appropriate to return an error there since we could be storing a modified session. Ugh. I think you're right.

**Reply by maxcountryman** [reply]
I can think of two immediate ways we might address this:

1. We add some kind of context to or around the record type to indicate if it's meant to be created and not already existing or
2. We follow along with the Django implementation and introduce a `create` method to the `SessionStore` trait; this would be called before saving when the session ID is `None`.

Because the record is currently directly serializable I'm not really sure where that context would be stored on the record itself. Maybe rather than encoding that directly on the type it could instead be a parameter, like `should`_create, passed to `save`. 

I don't really have a strong preference between the two (and maybe there better options I'm not immediately thinking of), both are breaking changes that require stores to update in some way. Even so, the failure mode is so catastrophic I feel like we must account for this.

**Comment by wt** [comment]
I think having a different method is important here to make it easier to
understand. I'm actually doing some experimentation to see what this could
look like. I have some ideas about incrementally changing so that it's not
an immediate major version bump.

Wren Turkal
You're more amazing than you think! ymatyt.com

**Reply by maxcountryman** [reply]
That makes sense to me.
